iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,87 +17,83 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
-#include <algorithm>
-#include <iterator>
-
 namespace iox
 {
 namespace roudi
 {
 cxx::expected<ServiceRegistry::Error> ServiceRegistry::add(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    // Forbid duplicate service descriptions entries
-    for (auto& element : m_serviceDescriptionVector)
+    auto index = find(serviceDescription);
+    if (index != NO_INDEX)
+    {
+        // entry exists, increment counter
+        auto& entry = m_serviceDescriptions[index];
+        entry->count++;
+        return cxx::success<>();
+    }
+
+    // entry does not exist, find a free slot if it exists
+
+    // fast path to a free slot (which was occupied by previously removed entry),
+    // prefer to fill entries close to the front
+    if (m_freeIndex != NO_INDEX)
```

Comment:
Actually one could improve this even further.
1. Introduce the fifo which can be found in `internal/concurrent/fifo.hpp` (its single push/single pop with an atomic) as `concurrent::FiFo<uint32_t, MAX_SERVICE_DESCRIPTIONS> m_freeIndices`
2. move all indices which are freed in there with just `m_freeIndices.push(indexToFree)`.
3. At this position we take a potential free index with `m_freeIndices.pop` and if it is not a nullopt use this one.

The solution above would not even increase the lines of code or would make the code more complex but would optimize it even further.

My personal opinion either keep it simple and make it work first before optimizing it but when one goes that route than take the low hanging fruits with you.

Reply:
Or we could use the index queue ... wouldn't than this service registry be thread safe in a lock free manner?! I have not really thought this through

Reply:
@elfenpiff I considered this, actually my `TypedAllocator` works like this. But here I opted against it, because I want that earlier slots in the vector are filled first.